### PR TITLE
Continuous rc merge

### PR DESCRIPTION
**Pull request by dtretiak**
Implementation of `CRCForecasterBase` and `CESNForecaster` and its corresponding training function. Developed in the `continuous_rc` branch and merge conflicts were resolved in `continuous_rc_merge` branch. 

Also removed mac and windows tests since usage is too high with them

### CODE REVIEW

File: src/orc/models/esn.py [hunk 1]
```diff
@@ -1,13 +1,14 @@
 """Classic ESN implementation with tanh nonlinearity and linear readout."""
```
Comment (Jan-Williams) [inline src/orc/models/esn.py]: Probably should update this to reflect continuous and discrete versions.

File: src/orc/models/esn.py [hunk 2]
```diff
@@ -147,6 +148,167 @@ def __init__(
...
+            out_dim=data_dim,
+            dtype=dtype,
+            seed=seed,
+            solver=solver,
+            stepsize_controller=stepsize_controller,
+        )
+        self.chunks = chunks
+
+
+def _solve_single_ridge_reg(res_seq, target_seq, beta):
```
Comment (Jan-Williams) [inline src/orc/models/esn.py:304]: I like the shift of ridge solve function out of `train_Forecaster`.

File: src/orc/rc.py [hunk 3]
```diff
@@ -167,3 +168,199 @@ def forecast_from_IC(self, fcast_len: int, spinup_data: Array) -> Array:
...
+        Stepsize controller to use for the ODE solver.
+
+    Methods
+    -------
+    force(in_seq, res_state)
+        Teacher forces the reservoir with the input sequence.
+    set_readout(readout)
+        Replaces the readout layer of the reservoir computer.
+    set_embedding(embedding)
+        Replaces the embedding layer of the reservoir computer.
```
Comment (Jan-Williams) [inline src/orc/rc.py:261]: should have `forecast` and `forecast_from_IC` here

Comment (Jan-Williams) [reply src/orc/rc.py:261]: Also in the `RCForecasterBase` docstring. I should've done that when I made the change from `ReservoirComputerBase` to `RCForecasterBase`.

File: src/orc/rc.py [hunk 4]
```diff
@@ -167,3 +168,199 @@ def forecast_from_IC(self, fcast_len: int, spinup_data: Array) -> Array:
...
+    @eqx.filter_jit
+    def forecast(self, ts: Array, res_state: Array) -> Array:
+        """Forecast from an initial reservoir state.
+
+        Parameters
+        ----------
+        ts : Array
+            Time steps for the forecast, (shape=(fcast_len,)).
+        res_state : Array
+            Initial reservoir state, (shape=(res_dim)).
```
Comment (Jan-Williams) [inline src/orc/rc.py]: should be (chunks, res_dim).

File: tests/models/test_esn.py
Comment (Jan-Williams) [inline tests/models/test_esn.py:1]: I think that we probably should use a fixture to get rid of some of the redundancy in the set up of the dummy test problems.

File: tests/models/test_esn.py [hunk 5]
```diff
@@ -136,3 +139,217 @@ def test_forecast_from_IC():
...
+    esn, R = orc.models.esn.train_CESNForecaster(
+        esn,
+        U_train,
+        ts_train,
+        initial_res_state=jax.numpy.zeros((chunks, res_dim), dtype=jnp.float64),
+    )
+    U_pred1 = esn.forecast(ts=ts_test, res_state=R[-1])
+    U_pred2 = esn.forecast_from_IC(ts=ts_test, spinup_data=U_train[-101:-1])
+
+    print("U_pred1 shape:", U_pred1.shape)
```
Comment (Jan-Williams) [inline tests/models/test_esn.py]: superfluous print statements?

**Review by Jan-Williams — COMMENTED** [review]
Super cool to see the continuous ESN working well with other components. Biggest overall comment is just that some docstrings haven't been updated as we've been updating the code. Finished the review so I'd have a place to keep track of those. Once I've addressed them, I'll hit approve.

Oh, also I wasn't quite sure what to make of the use of *args in the init for CRCForecasterBase. I think it could be useful to force explicitly including the args by name so that the hinting that pops up in whatever editor or IDE is more helpful when you try to inherit from it. Maybe it's not an issue and the hinting is smart enough to figure it out? I'll pull the continuous rc branch and check on this as I'm addressing the docstrings stuff.

File: src/orc/models/esn.py (hunk 2 above)
Comment (dtretiak) [reply src/orc/models/esn.py:304]: I'm liking the convention of leading underscores for helper functions. It makes it immediately apparent how those are used

File: src/orc/rc.py (hunk 3 above)
Comment (dtretiak) [reply src/orc/rc.py:261]: Ah both good catches thank you

File: tests/models/test_esn.py (hunk 5 above)
Comment (dtretiak) [reply tests/models/test_esn.py]: woops haha forgot to remove those after debugging, addressed in commit

File: tests/models/test_esn.py
Comment (dtretiak) [reply tests/models/test_esn.py:1]: agreed, let me know if you want me to tackle this

**Review by Jan-Williams — APPROVED** [review]
Finished up the changes to testing so all of the parallel setups are tested on the same problem defined in a fixture. Also checked on docstrings to make sure the shape annotations in the docs are correct. 

Approving now, but think we should chat about the notebook before actually merging.

**Comment by dtretiak** [comment]
Just ran the lorenz NB with the DT RC fixes. The plots show an off-by-one error on the forecast. 

Upon second thought, I'm wondering if it's worth including the point corresponding to the initial RC state? This would mean that (in the lorenz NB, for instance) we would have to include the final training datapoint into the test set. Maybe it's just cleaner to not output it and leave it as it was? I can then adjust the CT version to behave the same

**Comment by Jan-Williams** [comment]
> Just ran the lorenz NB with the DT RC fixes. The plots show an off-by-one error on the forecast.
> 
> Upon second thought, I'm wondering if it's worth including the point corresponding to the initial RC state? This would mean that (in the lorenz NB, for instance) we would have to include the final training datapoint into the test set. Maybe it's just cleaner to not output it and leave it as it was? I can then adjust the CT version to behave the same

I think that issue arises with the NB cause of  the line
`esn, R = orc.models.train_ESNForecaster(esn, U_train[-train_len-1:-1])`
So with the index change we talked about and changing that line to 
`esn, R = orc.models.train_ESNForecaster(esn, U_train[-train_len:])`
the off by one error you're talking about is fixed. IMO that's an argument for moving forward with the change, I think the second line here looks cleaner to me. I think that rather than include the final training datapoint into the test set, we can just force the entire sequence during training, but recognize that we have to truncate to remove the last hidden state before solving the regression.

**Comment by dtretiak** [comment]
Makes sense! I updated the CRC code to handle training and forecasting in the same way. 

Going forward, I think the cleanest way to setup the notebooks would be to just give the entire training sequence i.e. `esn, R = orc.models.train_ESNForecaster(esn, U_train)` and just adjust the train/test split to change the amount of training data. This can be saved for the NB updat PR though.

**Comment by Jan-Williams** [comment]
I agree deferring all notebook updates to NB update PRs. I'm happy to merge this now if you are.
